source/h2_decoder.c

```diff
@@ -15,48 +15,642 @@
+            bool succ = aws_byte_buf_write_from_whole_cursor(&decoder->scratch, to_read);
+            AWS_ASSERT(succ);
+            (void)succ;
+
+            /* If we have the correct number of bytes, call the state */
+            if (will_finish_state) {
+                const char *current_state = decoder->state.name;
+
+                struct aws_byte_cursor state_data = aws_byte_cursor_from_buf(&decoder->scratch);
+                err = decoder->state.fn(decoder, &state_data);
+
+                const size_t data_processed = decoder->scratch.len - state_data.len;
+                const size_t leftover_scratch = state_data.len;
+
+                if (leftover_scratch) {
+                    /* Move whatever unprocessed scratch we have back to the start of the buffer, and update the len */
+                    memmove(decoder->scratch.buffer, decoder->scratch.buffer + data_processed, leftover_scratch);
+                    decoder->scratch.len = leftover_scratch;
+                }
+
+                if (bytes_required > 0 && data_processed != bytes_required) {
+                    DECODER_LOGF(
+                        DEBUG,
+                        decoder,
+                        "Decoder state %s requested %" PRIu64 " bytes of data, but only used %" PRIu64,
+                        current_state,
+                        bytes_required,
+                        data_processed);
+                }
+            }
+        }
+
+        if (err) {
+            return err;
+        }
+    }
+
+    return AWS_OP_SUCCESS;
 }
 
 void aws_h2_decoder_set_logging_id(struct aws_h2_decoder *decoder, void *id) {
     decoder->logging_id = id;
 }
+
+/***********************************************************************************************************************
+ * State functions
+ **********************************************************************************************************************/
+
+static void s_decoder_set_state(struct aws_h2_decoder *decoder, const struct decoder_state *state) {
+    DECODER_LOGF(TRACE, decoder, "Moving from state %s to %s", decoder->state.name, state->name);
+
+    decoder->scratch.len = 0;
+    decoder->state = *state;
+}
+
+static void s_decoder_reset_state(struct aws_h2_decoder *decoder) {
+    s_decoder_set_state(decoder, &s_state_length);
+
+    DECODER_LOG(TRACE, decoder, "Resetting frame in progress");
+    AWS_ZERO_STRUCT(decoder->frame_in_progress);
+}
+
+static int s_state_fn_length(struct aws_h2_decoder *decoder, struct aws_byte_cursor *input) {
+
+    AWS_FATAL_ASSERT(input->len >= 3);
+
+    uint32_t payload_len = 0;
+    uint8_t *length_ptr = ((uint8_t *)&payload_len);
+
+    /* Read the first 3 bytes */
+    bool succ = aws_byte_cursor_read(input, length_ptr, 3);
+    AWS_ASSERT(succ);
+    (void)succ;
+
+    /* Reverse from network order */
+    payload_len = aws_ntoh24(payload_len);
+    /* Assert top byte isn't set */
+    AWS_FATAL_ASSERT((payload_len & 0xFF000000) == 0);
```

Comment:
Hi hello, please read above where I'm only reading 3 bytes. This assert is to validate ntoh24 works doesn't set the top byte. I can add that check for frame size, but since I don't support settings yet, it'll have to be hard coded.